src/tools/data_collection_util.cpp

```diff
@@ -0,0 +1,620 @@
+struct Input
+{
+public:
+  std::string dcName;
+
+  int uniformRefinements {0};
+  int polynomialOrder {2};
+
+  MeshForm meshForm {MeshForm::Box};
+
+  // Box mesh parameters
+  std::vector<double> boxMins;
+  std::vector<double> boxMaxs;
+  std::vector<int> boxResolution;
+  int boxDim {-1};
+
+  // File mesh parameters
+  std::string mfemFile;
+  std::vector<double> fileScale;
+  std::vector<double> fileTranslate;
+
+private:
+  bool m_verboseOutput {false};
+
+  /**
+   * Fixes up parameters associated with a box mesh based on user-provided input
+   */
+  void fixBoxParams()
+  {
+    const bool dimProvided = boxDim > 0;
+    const bool rangeProvided = !(boxMins.empty() && boxMaxs.empty());
+    const bool resProvided = !boxResolution.empty();
+
+    // First check: we need at least one of: dimension, mins/maxs, resolution
+    if(!(dimProvided || rangeProvided || resProvided))
+    {
+      throw "Box mesh must have at least one of: dimension, mins and maxs or "
+        "resolution to determine the dimension";
+    }
+
+    int szMins = boxMins.size();
+    int szMaxs = boxMaxs.size();
+    int szRes = boxResolution.size();
+
+    // Error checking on provided dimension
+    if(dimProvided)
+    {
+      if(boxDim < 1 || boxDim > 3)
+      {
+        throw "Dimension cannot exceed 3";
+      }
+
+      // Ensure that range and resolution have the right number of entries
+      if(rangeProvided && (szMins != boxDim || szMaxs != boxDim))
+      {
+        throw "Bounding box has different dimension than provided dimension";
+      }
+      if(resProvided && (szRes != boxDim))
+      {
+        throw "Box resolution has different dimension than provided dimension";
+      }
+    }
+
+    // Error checking on provided range; note: accounts for previous checks on dim
+    if(rangeProvided && (szMins != szMaxs))
+    {
+      throw "Bounding box mins and maxs has different dimensions";
+    }
+
+    // Error checking on provided range and resolution; note: accounts for previous checks on dim and range
+    if(rangeProvided && resProvided && (szMins != szRes))
+    {
+      throw "Bounding box mins and maxs has different dimensions than resolution";
+    }
+
+    // Now that we've checked errors, let's fill in missing data
+    const int dim = dimProvided ? boxDim : (resProvided ? szRes : szMins);
+    if(!dimProvided)
+    {
+      boxDim = dim;
+    }
+
+    if(!rangeProvided)  // set range to unit box
+    {
+      boxMins.resize(dim);
+      boxMaxs.resize(dim);
+      for(int i = 0; i < dim; ++i)
+      {
+        boxMins[i] = 0.;
+        boxMaxs[i] = 1.;
+      }
+    }
+
+    if(!resProvided)  // set resolution to 1
+    {
+      boxResolution.resize(dim);
+      for(int i = 0; i < dim; ++i)
+      {
+        boxResolution[i] = 1;
+      }
+    }
+  }
+
+public:
+  Input() = default;
+
+  bool isVerbose() const { return m_verboseOutput; }
+
+  void parse(int argc, char** argv, CLI::App& app)
```

Comment:
This might be a Maslow's hammer-type situation but to me this looks like a lot of command-line options; had you given any thought to using Inlet here?

Reply:
Thanks for the suggestion.
I considered it, but I think it would be overkill in this case.
There are actually two different modes, and each one only has 3-ish main options.
e.g.
```bash
> srun -n8 ./bin/data_collection_util -f file -m star-q2.mesh --scale 5
> srun -n8 ./bin/data_collection_util --min 0 0 --max 1 1 --res 64 64 -p 2
```

I split these up into cli's `option_group`s, which provides a better help/usage message:
```bash
>./bin/data_collection_util -h
Utility tool to create a blueprint compliant data store
Usage: ./bin/data_collection_util [OPTIONS]

Options:
  -h,--help                                       Print this help message and exit
  -o,--output-name TEXT                           Name of the output mesh. Defaults to box{2,3}d for box meshes, and the mfem mesh name when loading from an mfem file
  -v,--verbose                                    Enable/disable verbose output
  -l,--ref-level INT:NONNEGATIVE=0                The number of uniform refinement levels to apply to the mesh
  -f,--mesh-form ENUM:value in {box->0,file->1} OR {0,1}=0
                                                  Determines the input type -- either box or file
[Option Group: box]
  Options for setting up a box mesh
  Options:
    --min FLOAT x 2 Needs: --max                    Min bounds for box mesh (x,y[,z])
    --max FLOAT x 2 Needs: --min                    Max bounds for box mesh (x,y[,z])
    --res INT x 2                                   Resolution of the box mesh (i,j[,k])
    -d,--dimension INT:POSITIVE                     Dimension of the box mesh
    -p,--polynomial-order INT:NONNEGATIVE=2         polynomial order of the generated mesh
[Option Group: file]
  Options for loading from an mfem mesh file
  Options:
    -m,--mfem-file TEXT:FILE                        Path to a mesh file in the mfem format
    --scale FLOAT                                   Scale factors for the file mesh. Can either be one value or mesh.Dimension() values
    --translate FLOAT x 2                           Translation for the file mesh (sx,sy[,sz])
```